src/remmina.1
.\" Generated by scdoc 1.11.2
.\" Complete documentation for this program is not available as a GNU info page
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.\" Begin generated content:
.TH "REMMINA" "1" "2022-03-25"
.PP
.SH NAME
.PP
remmina -- Remmina the GTK+ Remote Desktop Client
.PP
.SH SYNOPSIS
.PP
remmina [-a|i|n|q|v] [-c FILE] [-e FILE] [-p TABINDEX] [-s SERVER] 
[-t PROTOCOL] [-x PLUGIN] [--update-profile] 
[--set-option OPTION[=VALUE]] [--display DISPLAY]
.PP
.SH DESCRIPTION
.PP
Remmina is a remote desktop client written in GTK+, aiming to be useful for
system administrators and travellers, who need to work with lots of remote
computers in front of either large monitors or tiny netbooks.\& Remmina
supports multiple network protocols in an integrated and consistent user
interface.\&
.PP
Currently RDP, VNC, SSH, SPICE, and WWW are supported.\&
.PP
Remmina is released in separated source packages:
.PP
.PD 0
.IP \(bu 4
"remmina", the main GTK+ application
.IP \(bu 4
"remmina-plugins", a set of plugins
.PD
.PP
Remmina is free and open-source software, released under GNU GPL license.\&
.PP
.SH FILES
.PP
${XDG_CONFIG_DIRS}/remmina.\&pref or ${XDG_CONFIG_HOME}/remmina/remmina.\&pref :
.PP
Remmina configuration files.\&
.PP
At the first Remmina execution the system wide Remmina configuration files,
will be copied in the ${XDG_CONFIG_HOME}
.PP
${XDG_DATA_DIRS}/FILE.\&remmina or ${XDG_DATA_HOME}/remmina/FILE.\&remmina :
.PP
Remmina profiles, the file name is autogenerated, but you can create
manually your own files with the FILE name you prefer.\& It'\&s possible to
specify a custom profile name and location.\&
.PP
.SH OPTIONS
.PP
\fB-h, --help\fR
.RS 4
Show help options
.PP
.RE
\fB-a\fR
.RS 4
Show about dialog
.PP
.RE
\fB-c, --connect=FILE\fR
.RS 4
Connect directly either to a desktop using options described in a file,
or a supported URI (RDP, VNC, SSH or SPICE).\&
.br
The filetype can be ".\&remmina" or one supported by a plugin capable of
importing files
.PP
.RE
\fB-e, --edit=FILE\fR
.RS 4
Open and edit desktop connection using options described by file,
file type can be either .\&remmina or one supported by a file
import capable plugin
.PP
.RE
\fB-k, --kiosk\fR
.RS 4
Start Remmina in kiosk mode (thin client)
.PP
.RE
\fB-n, --new\fR
.RS 4
Create a new connection profile
.PP
.RE
\fB-p, --pref=TABINDEX\fR
.RS 4
Show preferences dialog page
.PP
.RE
\fB-x, --plugin=PLUGIN\fR
.RS 4
Execute the plugin
.PP
.RE
\fB-q, --quit\fR
.RS 4
Quit the application
.PP
.RE
\fB-s, --server=SERVER\fR
.RS 4
Use default server name (for \fB--new\fR)
.PP
.RE
\fB-t, --protocol=PROTOCOL\fR
.RS 4
Use default protocol (for \fB--new\fR)
.PP
.RE
\fB-i, --icon\fR
.RS 4
Start as tray icon
.PP
.RE
\fB-v, --version\fR
.RS 4
Show the application'\&s version
.PP
.RE
\fB--update-profile\fR
.RS 4
Modify connection profile, require also \fB--set-option\fR
.PP
.RE
\fB--set-option OPTION[=VALUE]\fR
.RS 4
Set one or more profile settings, to be used with \fB--update-pro-file\fR
.PP
.RE
\fB--encrypt-password\fR
.RS 4
Encrypt a password
.PP
.RE
\fB--display=DISPLAY\fR
.RS 4
X display to use
.PP
.RE
\fB--disable-toolbar\fR
.RS 4
Disable toolbar
.PP
.RE
\fB--enable-fullscreen\fR
.RS 4
Enable fullscreen
.PP
.RE
\fB--enable-extra-hardening\fR
.RS 4
Enable extra hardening (disable closing confirmation, disable unsafe shortcut keys, hide tabs, hide search bar)
.br
List of disabled shortcut keys:
.br
fullscreen, autofit, prevtab and nexttab, scale, grab, minimize, viewonly, screenshot, disconnect, toolbar
.PP
.RE
\fB--no-tray-icon\fR
.RS 4
Disable tray icon
.PP
.RE
.SH EXAMPLES
.PP
To connect using an exisitng connection profile use:
\fBremmina -c FILE.\&remmina\fR
.PP
To quick connect using a URI:
.PP
remmina -c rdp://username@server
.br
remmina -c rdp://domain\\\\username@server
.br
remmina -c vnc://username@server
.br
remmina -c vnc://server?\&VncUsername=username
.br
remmina -c ssh://user@server
.br
remmina -c spice://server
.PP
To quick connect using a URI along with an encrypted password:
.PP
remmina -c rdp://username:encrypted-password@server
.br
remmina -c vnc://username:encrypted-password@server
.br
remmina -c vnc://server?\&VncUsername=username\\&VncPassword=encrypted-password
.PP
To encrypt a password for use with a URI:
.PP
remmina --encrypt-password
.PP
To update username and password and set a different resolution mode of a
remmina connection profile use:
.PP
echo "ausernamenapassword" | remmina --update-profile /PATH/TO/FOO.\&remmina --set-option username --set-option resolution_mode=2 --set-option password
.PP
.SH SEE ALSO
.PP
remmina-file-wrapper(1)
.PP
.SH AUTHORS
.PP
Antenore Gatta <antenore at simbiosi dot org> and
Giovanni Panozzo <giovanni at panozzo dot it>
.PP
See the THANKS file for a more detailed list.\&
.PP
Remmina was initially written by Vic Lee <llyzs@163.\&com>
.PP
This manual page was written by Antenore Gatta <antenore at simbiosi dot org>.\&
.PP
.SH COPYRIGHT
.PP
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2, or (at your option) any later
version.\&


ChangeLog
Don't edit anymore this file, use instead git@github.com:skywinder/github-changelog-generator.git to generate CHANGELOG.md

***** 1.0.0 - 2014-07-14 *****
Pull requests
o Typo in preference name broke key mapping in VNC
o Fix for issue #251:  RDP: No caret on explorer address bar and white squares
                            instead of small icons
o Fix issue with invisible toolbar in fullscreen.

o Fixes from repzilon and Mike Gilbert
  + repzilon
    - Replaced missed xzalloc with calloc
    - cmake: add WITH_TRANSLATIONS to disable .po files (Fixes: #23)
  + Mike Gilbert
    - Remove duplicate Actions key in desktop file
  + repzilon
    - Fixed Gentoo QA Notice in desktop file
  + repzilon
    - Recompressed icons. Also removed executable bit to some icons
  + repzilon
    - freerdp-channels was remed to freerdp-client
  + repzilon
    - plugins-rdp: fix crash when free'ing cursor
    - Fixes #120: Fix linking of avahi
    - cmake: small change in FindAVAHI
    - Fixes #113: call add_subdirectory after ALL add_definitions…
    - Detect GTK before other packages…
  + repzilon   Updated README with GTK2 on uclibc instructions
  + repzilon   Merge commit 'b6a55ae6f4633d55f8f03e7ce2eeb5899514a8fc'
  + repzilon   Corrected compile warnings
  + repzilon   GTK+2 compatibility
               * RDP mouse cursor image reference count is no longer decremented. It
                 probably introduces a memory leak, but decrementing the reference
                 count causes a segfault.
               * The Quick Connect bar of the main window no longer take half of the
                 window height.
               * The protocol selector of the Quick Connect bar looks like a regular
                 drop-down list on QtCurve theme.



***** 1.0.0 - 2012-02-10 *****
Main program design and functionality:
o Switch to the CMake build system
o Refactoring of code style
o Changing file name convention

RDP protocol related:
o Support for FreeRDP 1.0
o Support for RemoteFX
o Support for Network Level Authentication
o Added certificate validation prompt

***** 0.9.0 - 2010-12-12 *****
Main program design and functionality:
o Rewrite plugin system so that plugins can have more flexibilities
o Integrate scrolled window and fullscreen view modes into one button
o Remove quick connect menu and consolidate the feature with Connection->New
  menu. Move other Edit menu items under Connection menu as well. See
  http://bugs.debian.org/591755
o Change 'Quit' menu item to 'Close'
o Show version information in plugin window
o Add link button to resolution field to bring up to resolution preference
o Calculate auto-fit window size more correctly
o Add subgroup feature, using '/' as subgroup delimiter
o Remember group expansion status
o Add shortcut key names to toolbar tooltips
o Support shortcut key to be same as hostkey, which will be activated by just
  press(down and up) the hostkey. By default, this is the Keyboard grabbing
  shortcut.
o Add pin button to floating toolbar
o Add tray icon and various related preferences options. All featues in the
  original GNOME/XFCE panel applet have been ported in it, plus all windows
  can be minimized to tray (default is off)
o Add Help/Homepage menu

RDP protocol related:
o Support TLS negotiation added in FreeRDP 0.8
o Support audio input redirection added in FreeRDP 0.8
o Add new preference option 'Use client keyboard mapping' to obey X key
  mappings (wfaulk)

VNC protocol related:
o Change 'Proxy Destination' to 'Repeater' to make it clearer. Now the 'Server'
  is the server and 'Repeater' is the repeater. :)

NX protocol related:
o Add SSH tunnel

SSH protocol related:
o Add ability to spawn new session from toolbar menu
o Add Copy/Paste toolbar menu for SSH terminal
o Add Font and Scrollback lines preference option
o Add ability to show/hide hidden files in SFTP
o Remember paned position in SFTP window
o Add startup path option in SFTP

***** 0.8.0 - 2010-07-10 *****
Main program design and functionality:
o Introduce a new plugin system. All protocols except SSH are moved to a new
  source package 'remmina-plugins' so that they can have bug-fixes and releases
  separately. A new Plugins menu is added accordingly
o Add Hostkey feature and related Preference options. Hostkey feature provides
  an easy and reliable way to access various remote desktop toolbar functions
  through customizable shortcut key combinations
o Add Attachable / Detachable Tabs feature. Tabs now can be merged or separated
  between windows.
o Add ability to show / hide the toolbar through Hostkey
o Add Quick Search filter feature
o Add 'Save Settings When Starting the Connection' preference option, default
  is true. If set to false, profile will not be saved if only 'Connect' button
  is clicked
o Add 'Always show tab' option
o Add a little cosmetic motion effect to the floating toolbar
o Add Debug Window feature in Help menu
o Add Online Wiki link in Help menu
o Support IPv6 address notation
 
RDP protocol related:
o Bring in FreeRDP (http://freerdp.sourceforge.net) project as rdesktop
  replacement. Most of the features described below are direct benefits
  from the FreeRDP project
o FreeRDP provides a new keyboard system which is based on XKB and Windows
  keyboard layouts. In most cases it can detect the keyboard layout and work
  out of the box. It's also possible to manually assign a keyboard layout.
o Scaling remote screen is supported
o More clipboard formats (HTML, bitmaps, etc) are supported
o Better disk/printer sharing. Works with Windows 7 and Windows 2008
o Countless of original rdesktop bug-fixes and clean up
o Support importing Windows and tsclient .rdp file, either by drag-n-drop to
  the main window, by Tools / Import menu, or by -e / -c command line argument
o Support exporting Windows .rdp file

VNC protocol related:
o Support disabling TLS/VeNCrypt encryption when connecting to VNC servers
  that supports it, like Vino
o Support Refresh screen

src/remmina.c
/*
 * Remmina - The GTK Remote Desktop Client
 * Copyright (C) 2014-2022 Antenore Gatta, Giovanni Panozzo
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 *
 *  In addition, as a special exception, the copyright holders give
 *  permission to link the code of portions of this program with the
 *  OpenSSL library under certain conditions as described in each
 *  individual source file, and distribute linked combinations
 *  including the two.
 *  You must obey the GNU General Public License in all respects
 *  for all of the code used other than OpenSSL. *  If you modify
 *  file(s) with this exception, you may extend this exception to your
 *  version of the file(s), but you are not obligated to do so. *  If you
 *  do not wish to do so, delete this exception statement from your
 *  version. *  If you delete this exception statement from all source
 *  files in the program, then also delete it here.
 *
 */

#define G_LOG_USE_STRUCTURED
#ifndef G_LOG_DOMAIN
#define G_LOG_DOMAIN    ((gchar*)"remmina")
#endif  /* G_LOG_DOMAIN */
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#else
#include <gdk/gdkwayland.h>
#endif
#include <gio/gio.h>
#include <glib/gi18n.h>
#include <stdlib.h>

#include "config.h"
#include "remmina_sodium.h"
#include "remmina.h"
#include "remmina_exec.h"
#include "remmina_file_manager.h"
#include "remmina_icon.h"
#include "remmina_main.h"
#include "remmina_masterthread_exec.h"
#include "remmina_plugin_manager.h"
#include "remmina_plugin_native.h"
#ifdef WITH_PYTHONLIBS
#include "remmina_plugin_python.h"
#endif
#include "remmina_pref.h"
#include "remmina_public.h"
#include "remmina_sftp_plugin.h"
#include "remmina_ssh_plugin.h"
#include "remmina_widget_pool.h"
#include "remmina/remmina_trace_calls.h"

#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#include <pthread.h>
#ifdef HAVE_LIBGCRYPT
#include <gcrypt.h>
# if GCRYPT_VERSION_NUMBER < 0x010600
GCRY_THREAD_OPTION_PTHREAD_IMPL;
#endif  /* !GCRYPT_VERSION_NUMBER */
#endif  /* HAVE_LIBGCRYPT */

#ifdef HAVE_LIBGCRYPT
# if GCRYPT_VERSION_NUMBER < 0x010600
static int gcrypt_thread_initialized = 0;
#endif  /* !GCRYPT_VERSION_NUMBER */
#endif  /* HAVE_LIBGCRYPT */

gboolean kioskmode;
gboolean disabletoolbar;
gboolean fullscreen;
gboolean extrahardening;
gboolean disabletrayicon;

static GOptionEntry remmina_options[] =
{
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "about",	      'a',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Show \'About\'"),								     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "connect",	      'c',  0,			  G_OPTION_ARG_FILENAME_ARRAY, NULL, N_("Connect either to a desktop described in a file (.remmina or a filetype supported by a plugin) or a supported URI (RDP, VNC, SSH or SPICE)"),	     N_("FILE")	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ G_OPTION_REMAINING, '\0', 0,			  G_OPTION_ARG_FILENAME_ARRAY, NULL, N_("Connect to a desktop described in a file (.remmina or a filetype supported by a plugin)"),	     N_("FILE")	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "edit",	      'e',  0,			  G_OPTION_ARG_FILENAME_ARRAY, NULL, N_("Edit desktop connection described in file (.remmina or a filetype supported by plugin)"), N_("FILE")	},
	{ "help",	      '?',  G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,	       NULL, NULL,										     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "kiosk",	      'k',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Start in kiosk mode"),								     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "new",	      'n',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Create new connection profile"),						     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "pref",	      'p',  0,			  G_OPTION_ARG_STRING,	       NULL, N_("Show preferences"),								     N_("TABINDEX")	},
#if 0
	/* This option was used mainly for telepathy, let's keep it if we will need it in the future */
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	//{ "plugin",	      'x',  0,			  G_OPTION_ARG_STRING,	       NULL, N_("Run a plugin"),								     N_("PLUGIN")	},
#endif
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "quit",	      'q',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Quit"),									     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "server",	      's',  0,			  G_OPTION_ARG_STRING,	       NULL, N_("Use default server name (for --new)"),						     N_("SERVER")	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "protocol",	      't',  0,			  G_OPTION_ARG_STRING,	       NULL, N_("Use default protocol (for --new)"),						     N_("PROTOCOL") },
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "icon",	      'i',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Start in tray"),								     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "version",	      'v',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Show the application version"),						     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "full-version",     'V',  0,			  G_OPTION_ARG_NONE,	       NULL, N_("Show version of the application and its plugins"),				     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "update-profile",   0,    0,			  G_OPTION_ARG_FILENAME,       NULL, N_("Modify connection profile (requires --set-option)"),				     NULL	},
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	{ "set-option",	      0,    0,			  G_OPTION_ARG_STRING_ARRAY,   NULL, N_("Set one or more profile settings, to be used with --update-profile"),		     NULL	},
	{ "encrypt-password", 0,    0,			  G_OPTION_ARG_NONE,	       NULL, N_("Encrypt a password"),												  NULL		 },
	{ "disable-toolbar", 0,    0,			  G_OPTION_ARG_NONE,	       NULL, N_("Disable toolbar"),												  NULL		 },
	{ "enable-fullscreen", 0,    0,			  G_OPTION_ARG_NONE,	       NULL, N_("Enable fullscreen"),												  NULL		 },
	{ "enable-extra-hardening", 0,    0,		  G_OPTION_ARG_NONE,	       NULL, N_("Enable extra hardening (disable closing confirmation, disable unsafe shortcut keys, hide tabs, hide search bar)"),	  NULL		 },
	{ "no-tray-icon", 0,    0,			  G_OPTION_ARG_NONE,	       NULL, N_("Disable tray icon"),												  NULL		 },
	{ NULL }
};

#ifdef WITH_LIBGCRYPT
static int
_gpg_error_to_errno(gcry_error_t e)
{
	/* be lazy right now */
	if (e == GPG_ERR_NO_ERROR)
		return 0;
	else
		return EINVAL;
}
#endif /* !WITH_LIBGCRYPT */

static gint remmina_on_command_line(GApplication *app, GApplicationCommandLine *cmdline)
{
	TRACE_CALL(__func__);

	gint status = 0;
	gboolean executed = FALSE;
	GVariantDict *opts;
	gchar *str;
	const gchar **files;
	const gchar **remaining_args;
	gchar *protocol;
	gchar *server;

#if SODIUM_VERSION_INT >= 90200
	remmina_sodium_init();
#endif
	opts = g_application_command_line_get_options_dict(cmdline);

	if (g_variant_dict_lookup_value(opts, "disable-toolbar", NULL)) {
		disabletoolbar = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "enable-fullscreen", NULL)) {
		fullscreen = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "enable-extra-hardening", NULL)) {
		extrahardening = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "no-tray-icon", NULL)) {
		disabletrayicon = TRUE;
	}

	remmina_pref_init();

	if (g_variant_dict_lookup_value(opts, "quit", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_EXIT, NULL);
		executed = TRUE;
		status = 1;
	}

	if (g_variant_dict_lookup_value(opts, "about", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_ABOUT, NULL);
		executed = TRUE;
	}

	/** @warning To be used like -c FILE -c FILE -c FILE …
	 *
	 */
	if (g_variant_dict_lookup(opts, "connect", "^aay", &files)) {
		if (files)
			for (gint i = 0; files[i]; i++) {
				g_debug ("Connecting to: %s", files[i]);
				remmina_exec_command(REMMINA_COMMAND_CONNECT, files[i]);
			}
		executed = TRUE;
	}

	if (g_variant_dict_lookup(opts, G_OPTION_REMAINING, "^a&ay", &remaining_args)) {
		remmina_exec_command(REMMINA_COMMAND_CONNECT, remaining_args[0]);
		g_free(remaining_args);
		executed = TRUE;
	}

	if (g_variant_dict_lookup(opts, "edit", "^aay", &files)) {
		if (files)
			for (gint i = 0; files[i]; i++) {
				g_debug ("Editing file: %s", files[i]);
				remmina_exec_command(REMMINA_COMMAND_EDIT, files[i]);
			}
		//remmina_exec_command(REMMINA_COMMAND_EDIT, str);
		//g_free(str);
		executed = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "kiosk", NULL)) {
		kioskmode = TRUE;
		remmina_exec_command(REMMINA_COMMAND_MAIN, NULL);
		executed = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "new", NULL)) {
		if (!g_variant_dict_lookup(opts, "protocol", "&s", &protocol))
			protocol = NULL;

		if (g_variant_dict_lookup(opts, "server", "&s", &server))
			str = g_strdup_printf("%s,%s", protocol, server);
		else
			str = g_strdup(protocol);

		remmina_exec_command(REMMINA_COMMAND_NEW, str);
		g_free(str);
		executed = TRUE;
	}

	if (g_variant_dict_lookup(opts, "pref", "&s", &str)) {
		remmina_exec_command(REMMINA_COMMAND_PREF, str);
		executed = TRUE;
	}

	if (g_variant_dict_lookup(opts, "plugin", "&s", &str)) {
		remmina_exec_command(REMMINA_COMMAND_PLUGIN, str);
		executed = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "icon", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_NONE, NULL);
		executed = TRUE;
	}

	if (g_variant_dict_lookup_value(opts, "encrypt-password", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_ENCRYPT_PASSWORD, NULL);
		executed = TRUE;
		status = 1;
	}

	if (!executed)
		remmina_exec_command(REMMINA_COMMAND_MAIN, NULL);

	return status;
}

static void remmina_on_startup(GApplication *app)
{
	TRACE_CALL(__func__);

	RemminaSecretPlugin *secret_plugin;

	remmina_widget_pool_init();
	remmina_sftp_plugin_register();
	remmina_ssh_plugin_register();
	remmina_icon_init();

	g_set_application_name("Remmina");
	gtk_window_set_default_icon_name(REMMINA_APP_ID);

	/* Setting the X11 program class (WM_CLASS) is necessary to group
	* windows with .desktop file which has the same StartupWMClass */
	gdk_set_program_class(REMMINA_APP_ID);

	gtk_icon_theme_append_search_path(gtk_icon_theme_get_default(),
					  REMMINA_RUNTIME_DATADIR G_DIR_SEPARATOR_S "icons");
	g_application_hold(app);

	/* Check for secret plugin and service initialization and show console warnings if
	 * something is missing */
	secret_plugin = remmina_plugin_manager_get_secret_plugin();
	if (!secret_plugin)
		g_print("Warning: Remmina is running without a secret plugin. Passwords will be saved in a less secure way.\n");
	else
		if (!secret_plugin->is_service_available(secret_plugin))
			g_print("Warning: Remmina is running with a secrecy plugin, but it cannot connect to a secrecy service.\n");

	remmina_exec_command(REMMINA_COMMAND_AUTOSTART, NULL);
}

static gint remmina_on_local_cmdline(GApplication *app, GVariantDict *opts, gpointer user_data)
{
	TRACE_CALL(__func__);

	int status = -1;
	gchar *str;
	gchar **settings;

	/* Here you handle any command line options that you want to be executed
	 * in the local instance (the non-unique instance) */

	if (g_variant_dict_lookup_value(opts, "version", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_VERSION, NULL);
		status = 0;
	}

	if (g_variant_dict_lookup_value(opts, "full-version", NULL)) {
		remmina_exec_command(REMMINA_COMMAND_FULL_VERSION, NULL);
		status = 0;
	}

	if (g_variant_dict_lookup(opts, "update-profile", "^&ay", &str)) { /* ^&ay no need to free */
		if (g_variant_dict_lookup(opts, "set-option", "^a&s", &settings)) {
			if (settings != NULL) {
				status = remmina_exec_set_setting(str, settings);
				g_free(settings);
			} else {
				status = 1;
			}
		} else {
			status = 1;
			g_print("Error: --update-profile requires --set-option\n");
		}
	}

	/* Returning a non negative value here makes the application exit */
	return status;
}

int main(int argc, char *argv[])
{
	TRACE_CALL(__func__);
	GtkApplication *app;
	const gchar *app_id;
	int status;

	g_unsetenv("GDK_CORE_DEVICE_EVENTS");

	// Checking for environment variable "G_MESSAGES_DEBUG"
	// Give the less familiar with GLib a tip on where to get
	// more debugging info.
	if(!getenv("G_MESSAGES_DEBUG")) {
		/* TRANSLATORS:
		 * This link should point to a resource explaining how to get Remmina
		 * to log more verbose statements.
		 */
		g_message(_("Remmina does not log all output statements. "
			    "Turn on more verbose output by using "
			    "\"G_MESSAGES_DEBUG=all\" as an environment variable.\n"
			    "More info available on the Remmina wiki at:\n"
			    "https://gitlab.com/Remmina/Remmina/-/wikis/Usage/Remmina-debugging"
		));
	}

	/* Enable wayland backend only after GTK 3.22.27 or the clipboard
	 * will not work. See GTK bug 790031 */
	if (remmina_gtk_check_version(3, 22, 27))
		gdk_set_allowed_backends("wayland,x11,broadway,quartz,mir");
	else
		gdk_set_allowed_backends("x11,broadway,quartz,mir");

	remmina_masterthread_exec_save_main_thread_id();

	bindtextdomain(GETTEXT_PACKAGE, REMMINA_RUNTIME_LOCALEDIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);

#ifdef HAVE_LIBGCRYPT
# if GCRYPT_VERSION_NUMBER < 0x010600
	gcry_error_t e;
	if (!gcrypt_thread_initialized) {
		if ((e = gcry_control(GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread)) != GPG_ERR_NO_ERROR)
			return -1;
		gcrypt_thread_initialized++;
	}
#endif  /* !GCRYPT_VERSION_NUMBER */
	gcry_check_version(NULL);
	gcry_control(GCRYCTL_DISABLE_SECMEM, 0);
	gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
#endif  /* !HAVE_LIBGCRYPT */

	/* Initialize some Remmina parts needed also on a local instance for correct handle-local-options */
	remmina_pref_init();
	remmina_file_manager_init();

	remmina_plugin_manager_init();



	app_id = g_application_id_is_valid(REMMINA_APP_ID) ? REMMINA_APP_ID : NULL;
	app = gtk_application_new(app_id, G_APPLICATION_HANDLES_COMMAND_LINE | G_APPLICATION_CAN_OVERRIDE_APP_ID);
#if !GTK_CHECK_VERSION(4, 0, 0) /* This is not needed anymore starting from GTK 4 */
	g_set_prgname(app_id);
#endif
	g_application_add_main_option_entries(G_APPLICATION(app), remmina_options);
#if GLIB_CHECK_VERSION(2,56,0)
	gchar *summary = g_strdup_printf ("%s %s", app_id, VERSION);
	g_application_set_option_context_summary (G_APPLICATION(app), summary);
	g_free(summary);
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	g_application_set_option_context_parameter_string (G_APPLICATION(app), _("- or protocol://username:encryptedpassword@host:port"));
	// TRANSLATORS: Shown in terminal. Do not use characters that may be not supported on a terminal
	g_application_set_option_context_description (G_APPLICATION(app),
			_("Examples:\n"
				"To connect using an existing connection profile, use:\n"
				"\n"
				"\tremmina -c FILE.remmina\n"
				"\n"
				"To quick connect using a URI:\n"
				"\n"
				"\tremmina -c rdp://username@server\n"
				"\tremmina -c rdp://domain\\\\username@server\n"
				"\tremmina -c vnc://username@server\n"
				"\tremmina -c vnc://server?VncUsername=username\n"
				"\tremmina -c ssh://user@server\n"
				"\tremmina -c spice://server\n"
				"\n"
				"To quick connect using a URI along with an encrypted password:\n"
				"\n"
				"\tremmina -c rdp://username:encrypted-password@server\n"
				"\tremmina -c vnc://username:encrypted-password@server\n"
				"\tremmina -c vnc://server?VncUsername=username\\&VncPassword=encrypted-password\n"
				"\n"
				"To encrypt a password for use with a URI:\n"
				"\n"
				"\tremmina --encrypt-password\n"
				"\n"
				"To update username and password and set a different resolution mode of a Remmina connection profile, use:\n"
				"\n"
				"\techo \"username\\napassword\" | remmina --update-profile /PATH/TO/FOO.remmina --set-option username --set-option resolution_mode=2 --set-option password\n"));
#endif

	g_signal_connect(app, "startup", G_CALLBACK(remmina_on_startup), NULL);
	g_signal_connect(app, "command-line", G_CALLBACK(remmina_on_command_line), NULL);
	g_signal_connect(app, "handle-local-options", G_CALLBACK(remmina_on_local_cmdline), NULL);


	g_application_set_inactivity_timeout(G_APPLICATION(app), 10000);
	status = g_application_run(G_APPLICATION(app), argc, argv);
	g_object_unref(app);

	return status;
}
